lib/crx.test.js

```diff
@@ -0,0 +1,155 @@
+  // zip should be right after header.
+  return crxBuffer.subarray(HEADER_OFFSET + headerLength)
+}
+
+describe('generateCrx', () => {
+  const testDir = path.join('build', 'test-crx')
+  fs.mkdirSync(testDir, { recursive: true })
+
+  const contentPath = path.join(path.join(testDir, 'content'))
+  fs.mkdirSync(contentPath, { recursive: true })
+
+  const extensionKeyPath = path.join(testDir, 'extension-key.pem')
+  const publisherKeyPath = path.join(testDir, 'publisher-key.pem')
+  const publisherKeyAltPath = path.join(testDir, 'publisher-key-alt.pem')
+  const verifiedContentsKeyPath = path.join(testDir, 'verified-contents.pem')
+  let crxBuffer
+
+  before(async () => {
+    generatePrivateKey(extensionKeyPath)
+    generatePrivateKey(publisherKeyPath)
+    generatePrivateKey(publisherKeyAltPath)
+    generatePrivateKey(verifiedContentsKeyPath)
+
+    fs.writeFileSync(path.join(contentPath, 'manifest.json'), '{}')
+    fs.writeFileSync(path.join(contentPath, 'file1.js'), 'file1')
+    fs.writeFileSync(path.join(contentPath, 'file2.html'), 'file2')
+
+    const result = await crx.generateCrx(
+      contentPath,
+      extensionKeyPath,
+      [publisherKeyPath, publisherKeyAltPath],
+      verifiedContentsKeyPath
+    )
+    crxBuffer = result.crx
+  })
+
+  after(() => {
+    fs.removeSync(testDir, { recursive: true })
+  })
+
+  it('should have valid CRX header', () => {
+    assert.equal(crxBuffer.subarray(0, 4).toString(), 'Cr24') // Magic
+    assert.equal(crxBuffer.readUInt32LE(4), 3) // crx version
+  })
+
+  it('should include a valid ZIP payload', async () => {
+    const zip = await JSZip.loadAsync(getZipBuffer(crxBuffer))
+
+    const manifest = JSON.parse(await zip.files['manifest.json'].async('text'))
+    assert.equal(
+      manifest.update_url,
+      'https://clients2.google.com/service/update2/crx'
+    )
+
+    const id = util.getIDFromBase64PublicKey(manifest.key)
+    const header = getHeader(crxBuffer)
+    const signedData = crxpb.readSignedData(new Pbf(header.signed_header_data))
+    assert.equal(id, getIDFromBinary(signedData.crx_id))
+
+    assert.equal('file1', await zip.files['file1.js'].async('text'))
+    assert.equal('file2', await zip.files['file2.html'].async('text'))
+  })
+
+  it('should have valid verified contents', async () => {
+    const header = getHeader(crxBuffer)
+    const verifiedContents = JSON.parse(
+      zlib.gunzipSync(header.verified_contents)
+    )[0]
+    assert.equal(verifiedContents.description, 'treehash per file')
+    const payload = JSON.parse(
+      Buffer.from(
+        verifiedContents.signed_content.payload,
+        'base64url'
+      ).toString()
+    )
+    const signedData = crxpb.readSignedData(new Pbf(header.signed_header_data))
+    assert.equal(payload.item_id, getIDFromBinary(signedData.crx_id))
+    assert.equal(verifiedContents.signed_content.signatures.length, 1)
+  })
+
+  it('should have valid cryptographic signatures', async () => {
```

Comment:
needs a test for the failing case (invalid signature doesn't verify)

Reply:
Done.